docs: Deploy Loki Helm on AWS guide (#14517)
grafana/loki: docs/sources/setup/install/helm/deployment-guides/aws.md

{{< admonition type="tip" >}}
If you used the above EKSctl configuration file to create the cluster, you do not need to run this command. The OIDC provider is automatically installed.
{{< /admonition >}}

```bash
eksctl utils associate-iam-oidc-provider --cluster loki --approve
```

## Create S3 buckets

{{< admonition type="warning" >}}
 **DO NOT** use the default bucket names;  `chunk`, `ruler` and `admin`. Choose a **unique** name for each bucket. For more information see the following [security update](https://grafana.com/blog/2024/06/27/grafana-security-update-grafana-loki-and-unintended-data-write-attempts-to-amazon-s3-buckets/).
{{< /admonition >}}

Before deploying Loki, you need to create two S3 buckets; one to store logs (chunks), the second to store alert rules. You can create the bucket using the AWS Management Console or the AWS CLI. The bucket name must be globally unique.

{{< admonition type="note" >}}
GEL customers will require a third bucket to store the admin data. This bucket is not required for OSS users.
{{< /admonition >}}

```bash

> aws s3api create-bucket --bucket  < YOUR CHUNK BUCKET NAME eg. `loki-aws-dev-chunks`> --region <S3 region your account is on, eg `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 region your account is on, eg `eu-west-2`> \

Reviewer: ```suggestion
aws s3api create-bucket --bucket  <YOUR CHUNK BUCKET NAME e.g. `loki-aws-dev-chunks`> --region <S3 region your account is on, e.g. `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 region your account is on, e.g. `eu-west-2`> \
```

> aws s3api create-bucket --bucket  <YOUR RULER BUCKET NAME eg. `loki-aws-dev-ruler`> --region <S3 REGION your account is on, eg `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 REGION your account is on, eg `eu-west-2`>

Reviewer: ```suggestion
aws s3api create-bucket --bucket  <YOUR RULER BUCKET NAME e.g. `loki-aws-dev-ruler`> --region <S3 REGION your account is on, e.g. `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 REGION your account is on, e.g. `eu-west-2`>
```

    ```bash
    kubectl create namespace loki
    ```
### Loki Basic Authentication

Loki by default does not come with any authentication. Since we will be deploying Loki to AWS and exposing the gateway to the internet, we recommend adding at least basic authentication. In this guide we will give Loki a username and password:

1. To start we will need create a `.htpasswd` file with the username and password. You can use the `htpasswd` command to create the file:

   {{< admonition type="tip" >}}
    If you don't have the `htpasswd` command installed, you can install it using `brew` or `apt-get` or `yum` depending on your OS.
   {{< /admonition >}}

    ```bash
    htpasswd -c .htpasswd <username>
    ```
    This will create a file called `auth` with the username `loki`. You will be prompted to enter a password.

 1. Create a Kubernetes secret with the `.htpasswd` file:

    ```bash
    kubectl create secret generic loki-basic-auth --from-file=.htpasswd -n loki
    ```

    This will create a secret called `loki-basic-auth` in the `loki` namespace. We will reference this secret in the Loki Helm chart configuration.

    

### Loki Helm chart configuration

> Create a `values.yaml` file choosing the configuration options that best suit your requirements. Below there are two examples of `values.yaml` files for the Loki Helm chart. The first template is Loki in [microservices](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#microservices-mode) mode the second is Loki in [monolithic](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#monolithic-mode) mode with a replication factor of 3. The rest of this guide will focus on deploying Loki in microservices mode. However, there is no difference in the deployment process between the two modes apart from the components you are deploying.

Reviewer: ```suggestion
Create a `values.yaml` file choosing the configuration options that best suit your requirements. Below there is an example of `values.yaml` files for the Loki Helm chart in [microservices](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#microservices-mode) mode.
```

```yaml
loki:
   schemaConfig:
     configs:

>        - from: 2024-04-01

Reviewer: ```suggestion
       - from: "2024-04-01"
```

         store: tsdb
         object_store: s3
         schema: v13
         index:
           prefix: loki_index_
           period: 24h
   storage_config:
     aws:
       region: <S3 BUCKET REGION> # for example, eu-west-2  
       bucketnames: <CHUNK BUCKET NAME> # Your actual S3 bucket name, for example, loki-aws-dev-chunks
       s3forcepathstyle: false
   ingester:
       chunk_encoding: snappy
   pattern_ingester:
       enabled: true
   limits_config:
     allow_structured_metadata: true
     volume_enabled: true
     retention_period: 672h # 28 days retention
   compactor:
     retention_enabled: true 
     delete_request_store: s3
   ruler:
    enable_api: true
    storage:
      type: s3
      s3:

>         region: <Insert s3 bucket region> # for example, eu-west-2

Reviewer: Were you going to ALL CAP all of the placeholders?

- Full access to EKS (Amazon Elastic Kubernetes Service)
- Full access to S3 (Amazon Simple Storage Service)
- Sufficient permissions to create IAM roles and policies

There are two methods for authenticating and connecting Loki to AWS S3. We will guide you through the recommended method of granting access via an IAM role.

## Prerequisites

- Helm 3 or above. Refer to [Installing Helm](https://helm.sh/docs/intro/install/). This should be installed on your local machine.
- A running Kubernetes cluster on AWS. A simple way to get started is by using EKSctl. See [Getting started with EKSctl](https://eksctl.io/).
- Kubectl installed on your local machine. Refer to [Install and Set Up kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/).
- (Optional) AWS CLI installed on your local machine. Refer to [Installing the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html). This is required if you plan to use EKSctl to create the EKS cluster and modify the IAM roles and policies locally.
  
### EKS Minimum Requirements

{{< admonition type="caution" >}}

> These EKS requirements are the minimum specification needed to deploy Loki using this guide. You may wish to adjust plugins and instance types based on your AWS environment and workload. **If you choose to do so, this may invalidate the guide.**

Reviewer: Since you said this was a guide for BOTH monolithic and microservices, I think this admonition is going to be confusing for users.  Is this the minimum specification for monolithic?   OR for microservices?

    iam:
      withAddonPolicies:
        ebs: true
    volumeSize: 80
    volumeType: gp2
    ebsOptimized: true

```

The following plugins are also required to be installed within the EKS cluster:
- **Amazon EBS CSI Driver**: Enables Kubernetes to dynamically provision and manage EBS volumes as persistent storage for applications. We use this to provision the node volumes for Loki.
- **Amazon EKS Pod Identity Agent**: Manages AWS IAM roles for pods, allowing fine-grained access control to AWS resources without needing to store credentials in containers. This is how Loki will access the S3 bucket.
- **CoreDNS**: Provides internal DNS service for Kubernetes clusters, ensuring that services and pods can communicate with each other using DNS names. 
- **kube-proxy**: Maintains network rules on nodes, enabling communication between pods and services within the cluster.

You must also install an **OIDC (OpenID Connect) provider** on the EKS cluster. This is required for the IAM roles and policies to work correctly. If you are using EKSctl, you can install the OIDC provider using the following command:

```bash
eksctl utils associate-iam-oidc-provider --cluster loki --approve
```

> **This may be installed by default when creating the EKS cluster using EKSctl and the above config.**

Reviewer: Confused by this line.  Why is it bold?  Unclear on what "this" refers back to (OICD provider?).  Passive voice doesn't help either.  Are you trying to say that "The OICD provider was installed by default if you used the above config" ?

## Create three S3 buckets

{{< admonition type="warning" >}}
 **DO NOT** use the default bucket names;  `chunk`, `ruler` and `admin`. Choose a **unique** name for each bucket. For more information see the following [security update](https://grafana.com/blog/2024/06/27/grafana-security-update-grafana-loki-and-unintended-data-write-attempts-to-amazon-s3-buckets/).
{{< /admonition >}}

Before deploying Loki, you need to create two S3 buckets; one to store logs (chunks), the second to store alert rules. You can create the bucket using the AWS Management Console or the AWS CLI. The bucket name must be globally unique.

> {{<admonition type="note">}}

Reviewer: ```suggestion
{{< admonition type="note" >}}
```

GEL customers will require a third bucket to store the admin data. This bucket is not required for OSS users.

> {{</admonition>}}

Reviewer: ```suggestion
{{< /admonition >}}
```

```bash
aws s3api create-bucket --bucket  <Your S3 bucket eg. `loki-aws-dev-chunks`> --region <S3 region your account is on, eg `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 region your account is on, eg `eu-west-2`> \
aws s3api create-bucket --bucket  <Your S3 bucket eg. `loki-aws-dev-ruler`> --region <S3 region your account is on, eg `eu-west-2`> --create-bucket-configuration LocationConstraint=<S3 region your account is on, eg `eu-west-2`>
```

> Make sure to replace the region and bucket name with your desired values. We will revisit the bucket policy later in this guide.

Reviewer: ```suggestion
Make sure to replace the `region` and `bucket` name with your desired values. We will revisit the bucket policy later in this guide.
```

## Defining IAM roles and policies

The recommended method for connecting Loki to AWS S3 is to use an IAM role. This method is more secure than using access keys and secret keys which are directly stored in the Loki configuration. The role and policy can be created using the AWS CLI or the AWS Management Console. The below steps show how to create the role and policy using the AWS CLI.

{{< admonition type="tip" >}}

> Create a new directory and navigate to it. Make sure to create the files in this directory. All commands in this guide assume you are in this directory.

Reviewer: Why not just make this a step instead of a tip?

{{< /admonition >}}

1. Create a `loki-s3-policy.json` file with the following content:

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Sid": "LokiStorage",
                "Effect": "Allow",
                "Action": [
                    "s3:ListBucket",
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:DeleteObject"
                ],
                "Resource": [
                    "arn:aws:s3:::< Name of the chunk bucket >",
                    "arn:aws:s3:::< Name of the chunk bucket >/*",
                    "arn:aws:s3:::< Name of the ruler bucket >",
                    "arn:aws:s3:::< Name of the ruler bucket >/*"
                ]
            }
        ]
    }
    ```

>     **Make sure to replace the placeholder with the name of the buckets you created earlier.**

Reviewer: ```suggestion
    **Make sure to replace the placeholders with the names of the buckets you created earlier.**
```

    ```
    **Make sure to replace the placeholder with your AWS account ID.**

### Adding the policy to the S3 buckets

To allow the IAM role to access the S3 buckets, you need to add the policy to the bucket. You can do this using the AWS Management Console or the AWS CLI. The below steps show how to add the policy using the AWS CLI.

1. Create a bucket policy file named `bucket-policy-chunk.json` with the following content:

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Sid": "Statement1",
                "Effect": "Allow",
                "Principal": {
                    "AWS": "arn:aws:iam::<account id>:role/LokiServiceAccountRole"
                },
                "Action": [
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:DeleteObject",
                    "s3:ListBucket"
                ],
                "Resource": [
                    "arn:aws:s3:::<chunk bucket name>",
                    "arn:aws:s3:::<chunk bucket name>/*"
                ]
            }
        ]
    }
    ```

>     **Make sure to replace the placeholders with your AWS account ID and the bucket name.**

Reviewer: ```suggestion
    **Make sure to replace the placeholders with your AWS account ID and the bucket names.**
```

1. Add the policy to the bucket:

    ```bash
    aws s3api put-bucket-policy --bucket <Your S3 bucket eg. `loki-aws-dev-chunks`> --policy file://bucket-policy-chunk.json
    ```
1. Create a bucket policy file named `bucket-policy-ruler.json` with the following content:

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Sid": "Statement1",
                "Effect": "Allow",
                "Principal": {
                    "AWS": "arn:aws:iam::<account id>:role/LokiServiceAccountRole"
                },
                "Action": [
                    "s3:PutObject",
                    "s3:GetObject",
                    "s3:DeleteObject",
                    "s3:ListBucket"
                ],
                "Resource": [
                    "arn:aws:s3:::<ruler bucket name>",
                    "arn:aws:s3:::<ruler bucket name>/*"
                ]
            }
        ]
    }
    ```

>     **Make sure to replace the placeholders with your AWS account ID and the bucket name.**

Reviewer: ```suggestion
    **Make sure to replace the placeholders with your AWS account ID and the bucket names.**
```

1. Add the policy to the bucket:

    ```bash
    aws s3api put-bucket-policy --bucket <Your S3 bucket eg. `loki-aws-dev-ruler`> --policy file://bucket-policy-ruler.json
    ```

## Deploying the Helm chart

Before we can deploy the Loki Helm chart, we need to add the Grafana chart repository to Helm. This repository contains the Loki Helm chart.

1. Add the Grafana chart repository to Helm:

    ```bash
    helm repo add grafana https://grafana.github.io/helm-charts
    ```
1. Update the chart repository:

    ```bash
    helm repo update
    ```
1. Create a new namespace for Loki:

    ```bash
    kubectl create namespace loki
    ```
### Loki Basic Authentication

> Loki by default does not come with any authentication. Since we will be deploying Loki to AWS and exposing the gateway to the internet, we recommend adding atleast basic authentication. In this guide we will give Loki a username and password:

Reviewer: ```suggestion
Loki by default does not come with any authentication. Since we will be deploying Loki to AWS and exposing the gateway to the internet, we recommend adding at least basic authentication. In this guide we will give Loki a username and password:
```

    htpasswd -c .htpasswd <username>
    ```
    This will create a file called `auth` with the username `loki`. You will be prompted to enter a password.

 1. Create a Kubernetes secret with the `.htpasswd` file:

    ```bash
    kubectl create secret generic loki-basic-auth --from-file=.htpasswd -n loki
    ```

    This will create a secret called `loki-basic-auth` in the `loki` namespace. We will reference this secret in the Loki Helm chart configuration.

    

### Loki Helm chart configuration

Create a `values.yaml` file choosing the configuration options that best suit your requirements. Below there are two examples of `values.yaml` files for the Loki Helm chart. The first template is Loki in [microservices](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#microservices-mode) mode the second is Loki in [monolithic](https://grafana.com/docs/loki/<LOKI_VERSION>/get-started/deployment-modes/#monolithic-mode) mode with a replication factor of 3. The rest of this guide will focus on deploying Loki in microservices mode. However, there is no difference in the deployment process between the two modes apart from the components you are deploying.

> {{< collapse title="Distributed" >}}

Reviewer: Since you use the term "microservices" throughout the rest of the doc, I think it makes sense to use it here too.
```suggestion
{{< collapse title="Microservices" >}}
```

```yaml
 loki:
   schemaConfig:
     configs:
       - from: 2024-04-01
         store: tsdb
         object_store: s3
         schema: v13
         index:
           prefix: loki_index_
           period: 24h
   storage_config:
     aws:

>        region: <Insert s3 bucket region> # eu-west-2

Reviewer: ```suggestion
       region: <Insert s3 bucket region> # for example, eu-west-2
```

>        bucketnames: <Insert s3 bucket name> # Your actual S3 bucket name (loki-aws-dev-chunks)

Reviewer: ```suggestion
       bucketnames: <Insert s3 bucket name> # Your actual S3 bucket name, for example, loki-aws-dev-chunks
```

       s3forcepathstyle: false
   ingester:
       chunk_encoding: snappy
   pattern_ingester:
       enabled: true
   limits_config:
     allow_structured_metadata: true
     volume_enabled: true
     retention_period: 672h # 28 days retention
   compactor:
     retention_enabled: true 
     delete_request_store: s3
   ruler:
    enable_api: true
    storage:
      type: s3
      s3:

>         region: <Insert s3 bucket region> # eu-west-2

Reviewer: ```suggestion
        region: <Insert s3 bucket region> # for example, eu-west-2
```

>         bucketnames: <Insert s3 bucket name> # Your actual S3 bucket name (loki-aws-dev-ruler)

Reviewer: ```suggestion
        bucketnames: <Insert s3 bucket name> # Your actual S3 bucket name, for example, loki-aws-dev-ruler
```

  - The `ruler` section defines the configuration for the ruler, including the S3 bucket region and name. This tells Loki where to store the alert and recording rules.
  - For the full Loki configuration, refer to the [Loki Configuration]({{< relref "../../../../configure" >}}) documentation.

- **Storage:**
  - Defines where the Helm chart stores data.
  - Set the type to `s3` since we are using Amazon S3.
  - Configure the bucket names for the chunks and ruler to match the buckets created earlier.
  - The `s3` section specifies the region of the bucket.

- **Service Account:**
  - The `serviceAccount` section is used to define the IAM role for the Loki service account.
  - This is where the IAM role created earlier is linked.

- **Gateway:**
  - Defines how the Loki gateway will be exposed.
  - We are using a `LoadBalancer` service type in this configuration.

### Deploy Loki

Now that you have created the `values.yaml` file, you can deploy Loki using the Helm chart.

1. Deploy using the newly created `values.yaml` file:

    ```bash
    helm install --values values.yaml loki grafana/loki -n loki --create-namespace
    ```

>     **It is important to create a namespace called `loki` as our trust policy is set to allow the IAM role to be used by the `loki` service account in the `loki` namespace. This is configurable but make sure to update your service account**

Reviewer: ```suggestion
    **It is important to create a namespace called `loki` as our trust policy is set to allow the IAM role to be used by the `loki` service account in the `loki` namespace. This is configurable but make sure to update your service account.**
```

1. Verify the deployment:

    ```bash
    kubectl get pods -n loki
    ```
    You should see the Loki pods running.

>     ```console

Reviewer: You should clarify if this output is for the monolithic or microservices deployment (above).

{{< admonition type="caution" >}}
The Loki Gateway service is exposed to the internet. We provide basic authentication using a username and password in this tutorial. Refer to the [Authentication]({{< relref "../../../../operations/authentication" >}}) documentation for more information.
{{< /admonition >}}

To find the Loki Gateway service, run the following command:

```bash
kubectl get svc -n loki
```
You should see the Loki Gateway service with an external IP address. This is the address you will use to write to and query Loki.

```console
  NAME                             TYPE           CLUSTER-IP       EXTERNAL-IP                                                               PORT(S)              AGE
loki-gateway                     LoadBalancer   10.100.201.74    12345678975675456-1433434453245433545656563.eu-west-2.elb.amazonaws.com   80:30707/TCP         46m
```

Congratulations! You have successfully deployed Loki on AWS using the Helm chart. Before we finish, let's test the deployment.

## Testing Your Loki Deployment

> k6 is one of the fastest way to test your Loki deployment. This will allow you to both write and query logs to Loki. To get started with k6, follow the steps below:

Reviewer: ```suggestion
k6 is one of the fastest ways to test your Loki deployment. This will allow you to both write and query logs to Loki. To get started with k6, follow the steps below:
```
